statistics1-3

Mukhammed003 opened the pull request:
Выбранная командой 
          1. архитектура: MVVM;
          2. способ вёрстки: UIKit.
       
Ссылка на таск трекер: https://github.com/users/zanpakt0/projects/1/views/8

Сам файл .md с декомпозицией:
[statistics_Nurmukhanov.md](https://github.com/user-attachments/files/23728234/statistics_Nurmukhanov.md)

Первоначальная оценка времени выполнения задач и реальное время, потраченное на решении каждой задачи:

## Модуль 1:

#### Верстка экрана рейтинга
- Настройка navigationBar для кнопки сортировки (est: 1 час; fact: 21 минута).
- Таблица (est: 1 час; fact: 1 час 17 минут).
- Ячейка таблицы (est: 2 часа; fact: 1 час 28 минут).
- Кнопка сортировки (est: 1 час; fact: 49 минут).
- Сверстать Alert об ошибке загрузки данных (est: 1 час; fact: 35 минут).

#### Логика экрана рейтинга
- Настроить алерт сортировки (est: 1 час; fact: 45 минут).
- Сортировка (est: 2 часа; fact: 10 минут).
- Индикатор загрузки таблицы (est: 1 час; fact: 25 минут).
- Подстановка данных каждую ячейку таблицы (est: 2 часа; fact: 45 минут).
- Настроить взаимосвязи между model, view и view model для паттерна MVVM (est: 2 часа; fact: 1.5 часа).
- Логика подгрузки новых юзеров при скролле вниз (в таблице) (est: 2 часа; fact: 30 минут).
- Создание и настройка хранилища (UserDefaults) для хранения типа фильтра (est: 1 час; fact: 35 минут).
- Создание структур моделей (est: 1 час; fact: 23 минуты).

#### Данные в ячейках таблицы
- GET запрос аватара, имени и количества NFT для экрана рейтинга (est: 2 часа; fact: 2 часа 05 минут).

Mukhammed003 committed:
develope(newFeature): Создана структура моделей необходимая для работы с вкладкой Статистика

Выполнены следующие работы:
1) Добавлена структура User для декодирования ответа от сервера;
2) Добавлен класс UsersListCellViewModel для отображения данных в таблице с юзерами(только с необходимыми полями);

Mukhammed003 committed:
develope(newFeature): Настроены взаимосвязи между model, view и view model для паттерна MVVM

Выполнены следующие работы:
1) Добавлен view(StatisticsViewController) с параметром viewModel;
2) Добавлен view model(StatisticsViewModel) с параметром state(@Published) для дальнейшей связки;
2) Добавлен скелет bind(bindViewModel) для связки.

Mukhammed003 committed:
develope(newFeature): Настроен navigationBar для кнопки сортировки

Выполнены следующие работы:
1) В StatisticsViewController добавлена кнопка и настройка его как rightBarButtonItem;
2) Обновлён TabBarController с добавлением иконки для вкладки "Статистика".

Mukhammed003 committed:
develope(newFeature): Реализован метод по отправке запроса на получение списка юзеров

Выполнены следующие работы:
1) Обновлён NftService с добавлением метода по загрузке списка юзеров getUsers;
2) Обновлён StatisticsViewModel с добавлением метода по загрузке юзеров, и обновлении списка юзеров(cellViewModels);
3) При добавлении записей в cellViewModels производится проверка на дубликаты.

Mukhammed003 committed:
develope(newFeature): Сверстана ячейка таблицы по макету

Выполнены следующие работы:
1) Добавлен файл(StatisticsTableViewCell) для ячейки таблицы (для юзеров).

Mukhammed003 committed:
develope(newFeature): Сверстана таблица для юзеров

Выполнены следующие работы:
1) Добавлен метод configure в StatisticsTableViewCell для подстановки данных в каждую ячейку таблицы;
2) Сверстана таблица(tableViewWithUsers) в StatisticsViewController;
3) Реализованы необходимые методы для конфигурации таблицы с ячейками.

Mukhammed003 committed:
develope(newFeature): Реализована логика индикатора загрузки таблицы

Выполнены следующие работы:
1) Добавлен view для индикатора загрузки;
2) Обновлён bindViewModel для более корректного отображения нужных view в разных состояниях.

Mukhammed003 committed:
develope(newFeature): Реализована логика показа алерта об ошибке при получении списка юзеров с сервера

Выполнены следующие работы:
1) StatisticsViewController наследован от ErrorView, для реализации метода showError;
2) Реализован метод(showError) показа алерта об ошибке.

Mukhammed003 committed:
develope(newFeature): Реализована логика подгрузки новых юзеров при скролле вниз (в таблице с юзерами)

Выполнены следующие работы:
1) Реализован метод:
func tableView(_ tableView: UITableView, willDisplay cell: UITableViewCell, forRowAt indexPath: IndexPath) {}
для подгрузки новых юзеров при скролле вниз.

Mukhammed003 committed:
develope(newFeature): Сверстан алерт для сортировки

Выполнены следующие работы:
1) Добавлен алерт сортировки с двумя кнопками: "По имени", "По рейтингу". Пока без логики

Mukhammed003 committed:
develope(newFeature): Реализована логика фильтра

Выполнены следующие работы:
1) Добавлены методы для фильтра "По имени" и "По рейтинга";
2) Если ничего из фильтра не выбрано, дефолтно сортируется по По "Рейтингу";
3) Добавлено хранилище(UserDefaults) для хранения типа фильтра.

Mukhammed003 committed:
Добавил .md файл декомпозиции в корни проекта.

d00m1r commented on FakeNFT/Scenes /Statistics/NeedStructures/UserListCellViewModel.swift:
```diff
@@ -0,0 +1,33 @@
+    }
+    
+    static func == (lhs: UsersListCellViewModel, rhs: UsersListCellViewModel) -> Bool {
+        return lhs.name == rhs.name &&
```
В однострочных методах ключевое слово `return` можно опустить

d00m1r commented on FakeNFT/Scenes /Statistics/StatisticsViewController.swift:
```diff
@@ -0,0 +1,228 @@
+            self?.tableViewWithUsers.reloadData()
+        }
+        
+        self.showFilterActionSheet(firstAction: byNameAction, secondAction: byRatingAction, thirdAction: nil)
```
В коде присутствуют излишние обращения к `self.`

d00m1r commented on FakeNFT/Scenes /Statistics/StatisticsViewController.swift:
```diff
@@ -0,0 +1,228 @@
+            
+            tableViewWithUsers.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor, constant: 20),
+            tableViewWithUsers.leadingAnchor.constraint(equalTo: view.leadingAnchor, constant: 16),
+            tableViewWithUsers.trailingAnchor.constraint(equalTo: view.trailingAnchor, constant: -16),
```
Хорошей практикой является выносить константы в отдельный `enum`, для более простого изменения и чтения

d00m1r commented on FakeNFT/Scenes /Statistics/StatisticsViewController.swift:
```diff
@@ -0,0 +1,228 @@
+        let errorModel = ErrorModel(
+            message: "",
+            actionText: retryActionText,
+            action: {self.viewModel.fetchUsers(page: self.viewModel.pageNumber)}
```
`self` сильно захвачен, может приводить к RC

d00m1r reviewed (commented):
Рекомендую перепроверить код на соответствие [код-стайлу](https://habr.com/ru/articles/744470/)

Mukhammed003 committed:
fix(bugs): Поправил по замечаниям ревьюера после выполнения первого модуля

Выполнены следующие работы:
1) Добавлены константы для констрейнтов в отдельный enum;
2) Добавлены маркировки.
